internal/pkg/runtime/runtime.go

```diff
@@ -0,0 +1,44 @@
+package runtime
+
+import (
+	"fmt"
+	singularity "github.com/singularityware/singularity/internal/pkg/runtime/engine/singularity"
+	singularityConfig "github.com/singularityware/singularity/internal/pkg/runtime/engine/singularity/config"
+	runtime "github.com/singularityware/singularity/pkg/runtime"
+	"log"
+)
+
+var engines map[string]*runtime.RuntimeEngine
+
+// Instanciate a runtime engine based on json configuration
+func NewRuntimeEngine(name string, jsonConfig []byte) (*runtime.RuntimeEngine, error) {
+	var engine *runtime.RuntimeEngine
+
+	engine = engines[name]
+
+	if engine == nil {
+		return nil, fmt.Errorf("no runtime engine named %s found", name)
+	}
+	if err := engine.SetConfig(jsonConfig); err != nil {
+		return nil, fmt.Errorf("json parsing failed", err)
+	}
+	return engine, nil
+}
+
+// Register a runtime engine
+func registerRuntimeEngine(engine *runtime.RuntimeEngine, name string) {
+	if engines == nil {
+		engines = make(map[string]*runtime.RuntimeEngine)
+	}
+	engines[name] = engine
+	engine.RuntimeConfig = engine.InitConfig()
+	if engine.RuntimeConfig == nil {
+		log.Fatalf("failed to initialize %s engine\n", name)
+	}
+}
+
+func init() {
```

Comment:
I think each `RuntimeEngine` should have an init in its own `.go` file. For example, the OCI transports package has an example of how to pull off this registerEngine effect. See:

https://github.com/containers/image/blob/master/transports/transports.go For how they describe the registering of image transports
https://github.com/containers/image/blob/master/transports/alltransports/alltransports.go For a list of examples of image transports that implement this

Reply:
I took this way because my first thought was to embed all engines in one binary, so to not import all engines each time, I wrote a single init which register all engines resulting in one import for all engines. If we split engines into binaries, an init by engine is the right thing to do